### PR TITLE
[Moore] Fix mem2reg implementation

### LINKED ISSUE #7483: [Moore] Mem2Reg Error
Calling mem2reg on the following IR directly leads to `error: type mismatch for bb argument #0 of successor #0`.

```
  func.func private @"snitch_pkg::shared_offload"(%arg0: !moore.l32) -> !moore.l1 {
    %0 = moore.constant 0 : l32
    %offload = moore.variable : <l1>
    %false = hw.constant false
    cf.cond_br %false, ^bb13, ^bb1
  ^bb1:  // pred: ^bb0
    cf.cond_br %false, ^bb13, ^bb2
  ^bb2:  // pred: ^bb1
    cf.cond_br %false, ^bb13, ^bb3
  ^bb3:  // pred: ^bb2
    cf.cond_br %false, ^bb13, ^bb4
  ^bb4:  // pred: ^bb3
    cf.cond_br %false, ^bb13, ^bb5
  ^bb5:  // pred: ^bb4
    cf.cond_br %false, ^bb13, ^bb6
  ^bb6:  // pred: ^bb5
    cf.cond_br %false, ^bb13, ^bb7
  ^bb7:  // pred: ^bb6
    cf.cond_br %false, ^bb13, ^bb8
  ^bb8:  // pred: ^bb7
    cf.cond_br %false, ^bb13, ^bb9
  ^bb9:  // pred: ^bb8
    cf.cond_br %false, ^bb13, ^bb10
  ^bb10:  // pred: ^bb9
    cf.cond_br %false, ^bb13, ^bb11
  ^bb11:  // pred: ^bb10
    cf.cond_br %false, ^bb13, ^bb12
  ^bb12:  // pred: ^bb11
    cf.cond_br %false, ^bb13, ^bb14
  ^bb13:  // 13 preds: ^bb0, ^bb1, ^bb2, ^bb3, ^bb4, ^bb5, ^bb6, ^bb7, ^bb8, ^bb9, ^bb10, ^bb11, ^bb12
    cf.br ^bb15
  ^bb14:  // pred: ^bb12
    %1 = moore.conversion %0 : !moore.l32 -> !moore.l1
    moore.blocking_assign %offload, %1 : l1
    cf.br ^bb15
  ^bb15:  // 2 preds: ^bb13, ^bb14
    %2 = moore.read %offload : <l1>
    return %2 : !moore.l1
  }
```

### PR DESCRIPTION
**Pull request by fabianschuiki**
Fix a few issues in the mem2reg interface implementations of VariableOp, ReadOp, and BlockingAssignOp. Add tests reduced from the Snitch core that used to fail before this fix.

Fixes #7483.

### CODE REVIEW

**Comment by maerhart** [comment]
Fixes #7483 (I guess)

**Comment by fabianschuiki** [comment]
Yes exactly! Hadn't seen the issue yet 🙈

File: lib/Dialect/Moore/MooreOps.cpp [hunk 1]
```diff
@@ -334,6 +305,47 @@ LogicalResult VariableOp::canonicalize(VariableOp op,
   return success();
 }
 
+SmallVector<MemorySlot> VariableOp::getPromotableSlots() {
+  if (isa<SVModuleOp>(getOperation()->getParentOp()))
+    return {};
+  if (getInitial())
+    return {};
```
Comment (dobios) [inline lib/Dialect/Moore/MooreOps.cpp:312]: These two returns can be merged 
```suggestion
  if (isa<SVModuleOp>(getOperation()->getParentOp()) || getInitial())
    return {};
```

**Review by dobios — APPROVED** [review]
Just a small nit, but otherwise LGTM if Moore experts don't have anything against this!

**Review by hailongSun2000 — APPROVED** [review]
I don't have any questions.  LGTM :partying_face: !

**Pull request merged**